src/rigid_optimization/sdf_2_sdf_optimizer2d.cpp

```diff
@@ -0,0 +1,151 @@
+                                                           tsdf_generation_parameters.field_size,
+                                                           tsdf_generation_parameters.voxel_size,
+                                                           tsdf_generation_parameters.narrow_band_width_voxels);
+
+    eig::MatrixXf canonical_weight = canonical_field.replicate(1, 1);
+    for (int i=0; i < canonical_weight.rows(); ++i) { // Determine weight based on thickness
+        for (int j=0; j < canonical_weight.cols(); ++j){
+            canonical_weight(i, j) = (canonical_field(i, j) <= -eta) ? 0.0f : 1.0f;
+        }
+    }
+
+    eig::Vector3f twist = eig::Vector3f(0.0f, 0.0f, 0.0f);
+
+    for (int iteration_count=0; iteration_count<maximum_iteration_count; ++iteration_count) {
+        eig::Matrix3f matrix_A = eig::Matrix3f::Zero();
+        eig::Vector3f vector_b = eig::Vector3f::Zero();
+
+        eig::Matrix<float, 6, 1> twist3d;
+        twist3d << twist(0), 0.0f, twist(1), 0.0f, twist(2), 0.0f;
+        eig::Matrix4f twist_matrix3d = math::transformation_vector_to_matrix3d(twist3d);
+        eig::MatrixXf live_field = tsdf::generate_TSDF_2D(image_y_coordinate,
+                                                          live_depth_image,
+                                                          tsdf_generation_parameters.depth_unit_ratio,
+                                                          tsdf_generation_parameters.camera_intrinsic_matrix,
+                                                          tsdf_generation_parameters.camera_pose*twist_matrix3d,
+                                                          tsdf_generation_parameters.array_offset,
+                                                          tsdf_generation_parameters.field_size,
+                                                          tsdf_generation_parameters.voxel_size,
+                                                          tsdf_generation_parameters.narrow_band_width_voxels);
+        eig::MatrixXf live_weight = live_field.replicate(1, 1);
+        for (int i=0; i < live_weight.rows(); ++i) { // Determine weight based on thickness
+            for (int j=0; j < live_weight.cols(); ++j){
+                live_weight(i, j) = (live_field(i, j) <= -eta) ? 0.0f : 1.0f;
+            }
+        }
+        eig::Matrix<eig::Vector3f, eig::Dynamic, eig::Dynamic> live_gradient(canonical_field.rows(),
+                                                                             canonical_field.cols());
+        ropt::gradient_wrt_twist(live_field,
+                                 twist,
+                                 tsdf_generation_parameters.array_offset,
+                                 tsdf_generation_parameters.voxel_size,
+                                 live_gradient);
+
+        for (int i=0; i < live_field.rows(); ++i) {
+            for (int j=0; j < live_field.cols(); ++j) {
```

Comment:
Eigen matrices are column-major, so, for data locality, it makes better sense to traverse in a column-major order, i.e. rows would be the innter loop.

Reply:
fixed